include/envoy/registry/registry.h

```diff
@@ -172,6 +200,23 @@ template <class Base> class FactoryRegistry : public Logger::Loggable<Logger::Id
     }
   }
 
+  /**
+   * version is used for registering vendor specific factories that are versioned
+   * independently of Envoy.
+   */
+  static void registerFactory(Base& factory, absl::string_view name,
```

Comment:
Are we registering all the Envoy in-builts as well (using the semver of Envoy itself)? I think we should do this, as end users can optionally build Envoy with/without different filters, and it helps to know which ones are around.

Reply:
As far as I know, yes. There is RegisterInternalFactory that does not register a category, which is used for some internal factories (i.e. SslContext). I'm unsure if these need to be reported as well. Perhaps we should audit these.

Reply:
Can you pastebin what the full Node looks like somewhere? I'm curious to take a look at this.

Reply:
https://pastebin.com/KyZZYUty

Reply:
I notice that the build version information isn't populated; do we assume it's the same as the Envoy version? I think for semantic version this is fair. For labels, we actually have this metadata already in BUILD rules, see https://github.com/envoyproxy/envoy/blob/master/bazel/envoy_library.bzl#L70.

I wonder if we want to populate labels with the security posture and status of extensions? Seems a pretty cool use of them and should not be too hard.

Reply:
@markdroth will gRPC use SemVer for extension versioning? If not, what is your versioning scheme?
If we make an assumption that an empty extension version means that the version is the same as the "user agent", should the "user_agent_version" include SemVer as well? Will you version your xDS client using SemVer?

Reply:
The gRPC client itself does use semantic versioning, and it definitely makes sense to use SemVer as the format for user_agent_version.

The question of versions for plugins is an interesting one.  In principle, it seems reasonable for plugins to use their own semantic versioning, but in order to provide those versions to xDS, gRPC would probably need to provide a way for the plugin to record its version when it registers itself.  @ejona86 and @dfawley, thoughts...?

It might be a little confusing to assume that the default extension version matches the user agent version, because there could be some very simple plugins that don't bother to set a version.  It might make more sense to just treat an empty version as a unique value.

Reply:
I suppose how the empty version is treated is up to the xDS server. Are you suggesting removing the note about empty version from the proto doc?

Reply:
Yes.  I think we should just say that an empty version means no version specified, not that it defaults to the user agent version.